File: prototype/TAT-C/matlab/src/AbsoluteDate.m
classdef AbsoluteDate < handle
    % ABSOLUTEDATE: Utility for setting and converting between date formats
    %   Does not handle leap seconds
    
    properties (Access = private)
        %  Double.  The current date in Julian Day format
        currentDate = 2154500.0;
    end
    
    methods (Access = public)
        function obj = SetGregorianDate(obj,Year,Month,Day,Hour,Minute,Seconds)
            % Sets Gregorian date with input
            if Year < 1900 || Year > 2100
                error('Invalid Year provided to AbsoluteDate()');
            end
            if Month < 0 || Month > 12
                error('Invalid Month provided to AbsoluteDate()')
            end
            if Day < 0 || Day >= 31 % %TODO, check day by the month.
                error('Invalid Month provided to AbsoluteDate()')
            end
            if Hour < 0 || Hour >= 24
                error('Invalid Hour provided to AbsoluteDate()')
            end
            if Minute < 0 || Minute >= 60
                error('Invalid Hour provided to AbsoluteDate()')
            end
            if Seconds < 0 || Seconds >= 60
                error('Invalid Seconds provided to AbsoluteDate()')
            end
            obj.currentDate = obj.GregorianToJulianDate(Year,Month,Day,Hour,Minute,Seconds);
        end
        
        function obj = SetJulianDate(obj,julianDate)
            % Sets the Julian Date
            obj.currentDate =  julianDate;
        end
        
        function julianDate  = GetJulianDate(obj)
            % Sets the Julian Date
            julianDate = obj.currentDate;
        end
        
        function dateVec = GetGregorianDate(obj)
            % Returns the Gregorian date
            
            lMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
            jd = obj.currentDate;
            t1900 = (jd - 2415019.5) / 365.25;
            year = 1900 + fix(t1900);
            leapYrs = fix((year - 1900 - 1)*0.25);
            days = (jd - 2415019.5) - ((year - 1900) * 365 + leapYrs);
            
            % Handle leap year
            if days < 1
                year = year - 1;
                leapYrs = fix((year - 1900 - 1)*0.25);
                days = (jd - 2415019.5) - ((year - 1900) * 365 + leapYrs);
            end
            if mod(year, 4) == 0
                lMonth(2) = 29;
            end
            dayOfYr = fix(days);
            
            sumDays = 0;
            for i = 1:length(lMonth)
                sumDaysTemp = sumDays + lMonth(i);
                if sumDaysTemp >= dayOfYr
                    break;
                else
                    sumDays = sumDaysTemp;
                end
            end
            mon = i;
            
            day = dayOfYr - sumDays;
            tau = (days - dayOfYr) * 24;
            h = fix(tau);
            min = fix((tau - h) * 60);
            s = (tau - h - min/60) * 3600;
            
            dateVec = [year mon day h min s];
            
        end
        
        function Advance(obj,timeStepInSeconds)
            % Increments the time by the input number of seconds
            obj.currentDate = obj.currentDate + timeStepInSeconds/86400;
        end
        
    end
    
    methods (Access = private)
        
        function julianDate = GregorianToJulianDate(~,year,month,day,hour,min,sec)
            % Converts a Gregorian date to a Julian date
            jDay = 367 * year - fix(7 * (year + fix((month + 9) / 12)) / 4) + ...
                fix(275 * month / 9) + day + 1721013.5;
            partOfDay = ((sec/60 + min) / 60 + hour) / 24;
            julianDate = (jDay) + partOfDay;
        end
        
    end
end

File: prototype/TAT-C/cpp/src/AbsoluteDate.cpp
//------------------------------------------------------------------------------
//                           AbsoluteDate
//------------------------------------------------------------------------------
// GMAT: General Mission Analysis Tool.
//
// Copyright (c) 2002 - 2015 United States Government as represented by the
// Administrator of the National Aeronautics and Space Administration.
// All Other Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
// http://www.apache.org/licenses/LICENSE-2.0.
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied.   See the License for the specific language
// governing permissions and limitations under the License.
//
// Author: Wendy Shoan, NASA/GSFC
// Created: 2016.05.05
//
/**
 * Implementation of the AbsoluteDate class
 */
//------------------------------------------------------------------------------

#include "gmatdefs.hpp"
#include "DateUtil.hpp"
#include "AbsoluteDate.hpp"
#include "RealUtilities.hpp"
#include "GmatConstants.hpp"
#include "TATCException.hpp" // need new specific exception class?
#include "DateUtil.hpp"
#include "MessageInterface.hpp"

//#define DEBUG_GREGORIAN_DATE

//------------------------------------------------------------------------------
// static data
//------------------------------------------------------------------------------
const Integer AbsoluteDate::DAYS_PER_MONTH[12] =
   {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
   
const Real AbsoluteDate::JD_1900 = 2415019.5;

//------------------------------------------------------------------------------
// public methods
//------------------------------------------------------------------------------

//------------------------------------------------------------------------------
//  AbsoluteDate()
//------------------------------------------------------------------------------
/**
 * Default constructor
 */
//---------------------------------------------------------------------------
AbsoluteDate::AbsoluteDate() :
   currentDate         (GmatTimeConstants::JD_OF_J2000)
{
}

//------------------------------------------------------------------------------
//  AbsoluteDate(const AbsoluteDate &copy)
//------------------------------------------------------------------------------
/**
 * Copy constructor
 * 
 * @param copy the object to copy
 */
//---------------------------------------------------------------------------
AbsoluteDate::AbsoluteDate(const AbsoluteDate &copy) :
   currentDate         (copy.currentDate)
{
}

//------------------------------------------------------------------------------
//  AbsoluteDate & operator=(const AbsoluteDate &copy)
//------------------------------------------------------------------------------
/**
 * The operator= for the AbsolutDate class.
 * 
 * @param copy the object to copy
 */
//---------------------------------------------------------------------------
AbsoluteDate& AbsoluteDate::operator=(const AbsoluteDate &copy)
{
   if (&copy == this)
      return *this;
   
   currentDate         = copy.currentDate;
   
   return *this;
}

//------------------------------------------------------------------------------
//  ~AbsoluteDate()
//------------------------------------------------------------------------------
/**
 * Destructor
 */
//---------------------------------------------------------------------------
AbsoluteDate::~AbsoluteDate()
{
}

//------------------------------------------------------------------------------
//  void SetGregorianDate(Integer year, Integer month,  Integer day,
//                        Integer hour, Integer minute, Real second)
//------------------------------------------------------------------------------
/**
 * Sets the Gregorian date.
 * 
 * @param year   the year
 * @param month  the month
 * @param day    the day
 * @param hour   the hour
 * @param minute the minute
 * @param second the seconds 
 */
//---------------------------------------------------------------------------
void  AbsoluteDate::SetGregorianDate(Integer year, Integer month,  Integer day,
                                     Integer hour, Integer minute, Real second)
{
   // Sets Gregorian date from input
   if (year < 1900 || year > 2100)
      throw TATCException("Invalid Year provided to AbsoluteDate()\n");

   if (month < 0 || month > 12)
      throw TATCException("Invalid Month provided to AbsoluteDate()\n");

   bool isLeap = IsLeapYear(year); // DateUtil
   
   if (day < 0 || day >= DAYS_PER_MONTH[month-1])
      if (isLeap && month == 2 && day != 29)
         throw TATCException("Invalid Month provided to AbsoluteDate()\n");

   if (hour < 0 || hour >= 24)
      throw TATCException("Invalid Hour provided to AbsoluteDate()\n");

   if (minute < 0 || minute >= 60)
      throw TATCException("Invalid Hour provided to AbsoluteDate()\n");

   if (second < 0.0 || second >= 60.0)  // what about leap seconds??
      throw TATCException("Invalid Seconds provided to AbsoluteDate()\n");

   currentDate = GregorianToJulianDate(year,month,day,hour,minute,second);
}

//------------------------------------------------------------------------------
// void SetJulianDate(Real jd)
//------------------------------------------------------------------------------
/**
 * Sets the Julian date.
 * 
 * @param jd   the Julian date
 */
//---------------------------------------------------------------------------
void AbsoluteDate::SetJulianDate(Real jd)
{
   currentDate = jd;
}

//------------------------------------------------------------------------------
// Real GetJulianDate() const
//------------------------------------------------------------------------------
/**
 * Returns the Julian date.
 * 
 * @return   the Julian date
 */
//---------------------------------------------------------------------------
Real AbsoluteDate::GetJulianDate() const
{
   return currentDate;
}

//------------------------------------------------------------------------------
// Rvector6 GetGregorianDate()
//------------------------------------------------------------------------------
/**
 * Returns the Gregorian date.
 * 
 * @return   the Gregorian date
 */
//---------------------------------------------------------------------------
Rvector6 AbsoluteDate::GetGregorianDate()
{
   
   Integer year, month, day, hour, min;
   Real    second;
   Real    minute;
   
   Real    jd      = currentDate;
   Real    t1900   = (jd - JD_1900) / GmatTimeConstants::DAYS_PER_YEAR;
   year            = 1900 + GmatMathUtil::Fix(t1900);
   Integer leapYrs = GmatMathUtil::Fix((year - 1900 - 1)*0.25);
   Real days       = (jd - JD_1900) - ((year - 1900) * 365 + leapYrs);
   
   // Handle leap year
   if (days < 1)
   {
      year--;
      leapYrs = GmatMathUtil::Fix((year - 1900 - 1)*0.25);
      days    = (jd - JD_1900) - ((year - 1900) * 365 + leapYrs);
   }

   Real    dayOfYr    = GmatMathUtil::Fix(days);
   Integer dayOfYrInt = (Integer) dayOfYr;  
   
   ToMonthDayFromYearDOY(year, dayOfYrInt, month, day);  // DateUtil
#ifdef DEBUG_GREGORIAN_DATE
   MessageInterface::ShowMessage(
         "So far, year = %d, month = %d, day = %d, and (dayOfYr, dayOfYrInt,days) = (%12.10f %d %12.10f)\n",
         year, month, day, dayOfYr, dayOfYrInt, days);
#endif
   Real tau = (days - dayOfYr) * 24;
   hour     = GmatMathUtil::Fix(tau);
   minute   = GmatMathUtil::Fix((tau - hour) * 60);
   min      = (Integer) minute;
   second   = (tau - hour - minute/60) * GmatTimeConstants::SECS_PER_HOUR;
#ifdef DEBUG_GREGORIAN_DATE
   MessageInterface::ShowMessage(
         "   and then, tau = %12.10f, hour = %d, min = %d, second = %16.14f\n",
         tau, hour, min, second);
#endif
   
   Rvector6 result(year, month, day, hour, min, second);
   return result;
}

//------------------------------------------------------------------------------
// void Advance(Real stepInSec)
//------------------------------------------------------------------------------
/**
 * Advances the date by the number of seconds specified.
 * 
 * @param  stepInSec   stepsize (seconds)
 */
//---------------------------------------------------------------------------
void AbsoluteDate::Advance(Real stepInSec)
{
   currentDate += stepInSec / GmatTimeConstants::SECS_PER_DAY;
}

//------------------------------------------------------------------------------
// protected methods
//------------------------------------------------------------------------------

//------------------------------------------------------------------------------
// Real GregorianToJulianDate(Integer year, Integer month,  Integer day,
//                            Integer hour, Integer minute, Real second)
//------------------------------------------------------------------------------
/**
 * Converts a Gregorian date to a Julian date.
 * 
 * @param year   the year
 * @param month  the month
 * @param day    the day
 * @param hour   the hour
 * @param minute the minute
 * @param second the seconds 
 * 
 * @return the Julian date
 */
//---------------------------------------------------------------------------
Real AbsoluteDate::GregorianToJulianDate(Integer year, Integer month,  Integer day,
                                         Integer hour, Integer minute, Real second)
{
//   // Converts a Gregorian date to a Julian date
//   Real jDay = 367 * year - GmatMathUtil::Fix(7 *
//               (year + GmatMathUtil::Fix((month + 9) / 12)) / 4) +
//               GmatMathUtil::Fix(275 * month / 9) + day + 1721013.5;
//   Real partOfDay = ((second/60 + minute) / 60 + hour) / 24;
//   Real julianDate = (jDay) + partOfDay;
//   return julianDate;
   return JulianDate(year, month, day, hour, minute, second);
}
